File: android/src/main/java/com/salvedb/SyncTriggerDebouncer.kt
package com.salvedb

/** Coalesces rapid [signal] calls into at most one [onFire] per [windowMillis]. */
class SyncTriggerDebouncer(
  private val windowMillis: Long,
  private val nowMillis: () -> Long = System::currentTimeMillis,
  private val onFire: () -> Unit,
) {
  private var lastFiredAt: Long? = null

  @Synchronized
  fun signal() {
    val now = nowMillis()
    val last = lastFiredAt
    if (last != null && now - last < windowMillis) return

    lastFiredAt = now
    onFire()
  }
}

File: ios/Sync/SyncTriggerDebouncer.swift
import Foundation

/// Coalesces rapid `signal()` calls into at most one `onFire` per `windowSeconds`.
final class SyncTriggerDebouncer {
  private let windowSeconds: TimeInterval
  private let now: () -> TimeInterval
  private let onFire: () -> Void
  private var lastFiredAt: TimeInterval?

  init(
    windowSeconds: TimeInterval,
    now: @escaping () -> TimeInterval = { Date().timeIntervalSince1970 },
    onFire: @escaping () -> Void
  ) {
    self.windowSeconds = windowSeconds
    self.now = now
    self.onFire = onFire
  }

  func signal() {
    let current = now()
    if let last = lastFiredAt, current - last < windowSeconds {
      return
    }
    lastFiredAt = current
    onFire()
  }
}
